Add option to block zoom

LeSnake04 opened the pull request:
This is the result of my work so far.  It blocks scrolling unless done by the top bar (for now by default). The option is shown, but It doesn't toggle the setting yet.  

The toggle button is shown in the toolspage at the bottom (The Icon is a placeholder).

#539 
closes #543

LeSnake04 commented:
The error I got stuck at:
````
error[E0521]: borrowed data escapes outside of method
   --> rnote-ui/src/canvaswrapper.rs:245:17
    |
240 |           fn setup_input(&self) {
    |                          -----
    |                          |
    |                          `self` is a reference that is only valid in the method body
    |                          let's call the lifetime of this reference `'1`
...
245 | /                 self.canvas_zoom_scroll_controller.connect_scroll(clone!(@weak inst as canvaswrapper => @default-return Inhibit(false), move |controller, _, dy| {
246 | |                     if controller.current_event_state() == gdk::ModifierType::CONTROL_MASK {
247 | |                         if restrict_zoom() {
248 | |                             let new_zoom = canvaswrapper.canvas().engine().borrow().camera.total_zoom() * (1.0 - dy * RnCanvas::ZOOM_STEP);
...   |
259 | |                     }
260 | |                 }));
    | |                   ^
    | |                   |
    | |___________________`self` escapes the method body here
    |                     argument requires that `'1` must outlive `'static`
```

I coulnt find a way to tranfer the bool into the closure. Maybe someone knows a trick that could help here. 

It needs to grab the current state.

LeSnake04 commented:
My current idea would be kinda hacky: 
Set and read an environment variable that stores the state of the option. 

Is there a better way to solve this?

Kneemund commented:
Are you trying to disable all zooming functionality, or would disabling the zoom gesture suffice? The issue seems to refer to disabling the gesture, and there's already `RnCanvasWrapper#canvas_zoom_gesture_enable()`, which is used to disable the zoom gesture when the touch drawing mode is enabled. (Although I'm not sure if it also disables panning ATM.)

LeSnake04 commented:
> Are you trying to disable all zooming functionality, or would disabling the zoom gesture suffice? #543 seems to refer to disabling the gesture, and there's already [`RnCanvasWrapper#canvas_zoom_gesture_enable()`](https://github.com/flxzt/rnote/blob/main/rnote-ui/src/canvaswrapper.rs#L610-L621), which is used to disable the zoom gesture when the touch drawing mode is enabled. (Although I'm not sure if it also disables panning ATM, but I don't _think_ so.)

Its intended to be part of #539 The planned zoom tool will work when zoom is restricted making it much easier to zoom with this enabled.

And I think `ctrl+scroll` is something  you  could do out of reflexes, especially when you are used to moving/zooming with mouse and keyboard .

LeSnake04 commented:
About `canvas_gesture_zoom()`. This might  work. I will def try it.

LeSnake04 commented:
Update: `canvas_gesture_zom_enable` worked for pinch to zoom. 

For the other zoom methods, I found a way to get the property. Though for some reason it fails to set the property value. I added some dbg!(), and it looks the value given to set_property is correct regarding setting the property, but I still always get false when reading the property. It makes no sense to me...

LeSnake04 commented:
@flxzt can you please take a look?

LeSnake04 commented on rnote-ui/src/canvaswrapper.rs:
```diff
@@ -534,6 +557,16 @@ impl RnCanvasWrapper {
     }
+    #[allow(unused)]
+    pub(crate) fn restrict_zoom(&self) -> bool {
+        dbg!(self.property::<bool>("restrict-zoom"))
```
This always returns false, no matter what the state of the toggle button is.

flxzt commented:
Thanks for the PR! @Kneemund maybe you want to? Pretty busy this week, it might take a few days otherwise

Kneemund commented on rnote-ui/src/canvaswrapper.rs:
```diff
@@ -179,20 +181,30 @@ mod imp {
         fn property(&self, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
             match pspec.name() {
                 "show-scrollbars" => self.show_scrollbars.get().to_value(),
+                "restrict-zoom" => self.show_scrollbars.get().to_value(),
```
The issue lies here - you forgot to change `self.show_scrollbars` to `self.restrict_zoom`. Because of that, `self.property::<bool>("restrict-zoom")` will always return the current value of the scrollbar visibility option instead.

LeSnake04 replied:
Yes, that was it.

flxzt commented:
just one thing before implement too much: I think it should be an action entry in the canvas menu, not a button in the settings.

Kneemund commented:
My two cents on the UI:

While working on #490, I've also experimented with the UI and this was what I came up with:
(I haven't spent any time adding functionality to the toggle; especially the interaction between touch drawing and this option will require a bit more effort. I'll only implement the "zoom to real size" button then.)

![grafik](https://user-images.githubusercontent.com/49598842/224566211-f7c1056e-0a34-4ce6-a121-d9a12c6722d4.png)

An action entry like flxzt suggested would be smaller and more consistent with the other options though.

LeSnake04 commented:
@Kneemund i was thinking about putting it there as well. (Alongside the current location)

PhilDevProg commented:
> My two cents on the UI:
> 
> While working on #490, I've also experimented with the UI and this was what I came up with:
> (I haven't spent any time adding functionality to the toggle; especially the interaction between touch drawing and this option will require a bit more effort. I'll only implement the "zoom to real size" button then.)
> 
> ![grafik](https://user-images.githubusercontent.com/49598842/224566211-f7c1056e-0a34-4ce6-a121-d9a12c6722d4.png)
> 
> An action entry would be smaller and more consistent with the other options though.

I think that's the best looking and understandable solution. Maybe the other options (Pen Sounds/Document Borders/Touch Input) could work and look the same to be more consistent with the settings in the settings menu but that's something which should be discussed somewhere else. I think either do all options in this style with the toggle or do all in the current style with the checkmark.

LeSnake04 commented:
I am already working on it. Should be done soon

LeSnake04 commented:
In fact, I already created a working toggle in the style of the other buttons of the menu. 

The question is, which zooms should be blocked?
Pint to zoom should definitely be blocked, but I am not sure about the other zooms. I disabled all but the controls in the canvasmenu for now. I personally think Ctrl+scroll should be blocked as well, and maybe alt+shift+drag,but I would like your opinion.

LeSnake04 commented:
It should now compile and run. Its a checkbox toggle in the canvasmenu.

As far as I understand we now just have to decide 2 things:

- [ ] name of the button
- [ ] which ways of zooming should be blocked

Kneemund commented:
I think that touch zooming is the only way to accidentally zoom, so - assuming this option is only supposed to prevent accidental zooming - the gesture should be the only thing that gets disabled (and the option should be named accordingly). I can't think of a reason why you'd want to disable other zoom methods.

LeSnake04 commented:
Done.
Looks ready to merge from my side

Kneemund commented on rnote-ui/src/canvaswrapper.rs:
```diff
@@ -534,6 +551,16 @@ impl RnCanvasWrapper {
+    #[allow(unused)]
+    pub(crate) fn set_block_pinch_zoom(&self, block_pinch_zoom: bool) {
+        self.set_property("block-pinch-zoom", block_pinch_zoom);
+        self.canvas_zoom_gesture_enable(!block_pinch_zoom);
```
This causes conflicts with touch drawing, because that toggles the zoom gesture too. One example would be: touch drawing on & block zoom on, disable touch drawing -> gesture is enabled even though block zoom is on.

LeSnake04 replied:
Should be solved now

LeSnake04 commented:
I took the liberty to replace the canvas_zoom_gesture_enable function with an update one that sets the value based on block_pinch_zoom and touch_drawing.  I replaced it bc all calls were the same after my fix, so I made a helper function. There was no other use of the enable one aside the helper function, so I just replaced it.

I you want to keep the enable function for future usage, I can add it back in.

Kneemund reviewed (commented):
In terms of functionality, this LGTM. I only noticed some minor stylistic things I'd personally do differently (e.g. changing the condition of `canvas_zoom_gesture_update` to `!block_pinch_zoom && !touch_drawing`), but I'd rather leave that for @flxzt to review.

LeSnake04 commented:
I assume you mean || not &&

Kneemund commented:
No, `!(self.block_pinch_zoom() || self.canvas().touch_drawing())` is identical to `!self.block_pinch_zoom() && !self.canvas().touch_drawing()` ([De Morgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws)).

LeSnake04 commented:
Ty for the  rabbit hole 

> No, `!(self.block_pinch_zoom() || self.canvas().touch_drawing())` is identical to `!self.block_pinch_zoom() && !self.canvas().touch_drawing()` ([De Morgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws)).

Thanks for the rabbit hole. That will be an interesting dive for later.

flxzt commented on rnote-ui/src/appwindow/mod.rs:
```diff
@@ -610,6 +610,11 @@ impl RnAppWindow {
+            self,
+            "block-pinch-zoom",
+            &active_tab.block_pinch_zoom().to_variant(),
+        );
```
we could do it like this and it works, because the property `block-pinch-zoom` is updated here when the active tab changes. But I would actually prefer it if this would be a property binding, similar to the `touch-drawing` property.

Meaning:  
- also add a "global" `block-pinch-zoom` property on the appwindow
- create a binding to the canvaswrapper property in `canvaswrapper.init_reconnect()`
- refactor the action so that is a property-action of the appwindow `block-pinch-zoom` property.

flxzt commented:
I added some comments.

When those are addressed, there is only one thing missing: a boolean KV-pair in the settings, so that it is remembered across sessions. Check out `touch-drawing` how to create a settings bind for it.

LeSnake04 commented on rnote-ui/src/appwindow/appwindowactions.rs:
```diff
@@ -302,18 +302,6 @@ impl RnAppWindow {
-                action_block_pinch_zoom.set_state(block_pinch_zoom.to_variant());
-            }),
-        );
-
```
I removed it since I am not sure what it does and if its nessary for PropertyActions. I found no similar code for touch-drawing

LeSnake04 replied:
It seems to work without it.

flxzt replied:
If it's not overwritten by a custom callback, the default changes the action state (so: the property) in the same way

LeSnake04 commented:
I tried to make this a propertyaction by making it more like touch-drawing

But it panics with this message:
'Source property block-pinch-zoom on type RnAppWindow not found

Kneemund commented:
You haven't created the property in `RnAppWindow` yet, so you can't bind anything to it. Take a look at this:

https://github.com/flxzt/rnote/blob/228ab134d26c8f466ea9a21562a4c2991a2e9f2c/rnote-ui/src/appwindow/imp.rs#L151-L153

and all the other related properties and functions in `appwindow/imp.rs` and `appwindow/mod.rs`. You haven't done that for `block-pinch-zoom` yet.

LeSnake04 commented:
Thanks, that fixed it. 
@flxzt I added the changes you requested.

flxzt commented:
I forgot to mention that the property bind also needs to be unbound in canvaswrapper's `disconnect_handlers()`.

And I would move the action in the menu to below "Draw with Touch Input", it looks a bit odd where it is right now.

Other than that, this looks good to me. @Kneemund do you have anything else?

LeSnake04 commented:
Do you mean this?
https://github.com/flxzt/rnote/blob/ce512ffc0f5c24d90aed1e1622d5219c702c30d7/rnote-ui/src/canvaswrapper.rs#L570-L576

flxzt commented:
no, [here](https://github.com/flxzt/rnote/blob/228ab134d26c8f466ea9a21562a4c2991a2e9f2c/rnote-ui/src/canvaswrapper.rs#L548)

Kneemund commented:
> Other than that, this looks good to me. @Kneemund do you have anything else?

Looks good to me too.

LeSnake04 commented:
> I forgot to mention that the property bind also needs to be unbound in canvaswrapper's `disconnect_handlers()`.
> 
> And I would move the action in the menu to below "Draw with Touch Input", it looks a bit odd where it is right now.
> 
> Other than that, this looks good to me. @Kneemund do you have anything else?

Done
